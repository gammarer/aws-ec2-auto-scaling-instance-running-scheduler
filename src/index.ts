import * as crypto from 'crypto';
import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as scheduler from 'aws-cdk-lib/aws-scheduler';
import { Construct } from 'constructs';

export interface ScheduleProperty {
  readonly timezone: string;
  readonly minute?: string;
  readonly hour?: string;
  readonly week?: string;
}

export interface AutoScalingGroupsProperty {
  readonly groupName: string;
  readonly runningDesiredCapacity: number;
  readonly stopSchedule: ScheduleProperty;
  readonly startSchedule: ScheduleProperty;
}

export interface Ec2AutoScalingInstanceRunningSchedulerProps {
  readonly targets: AutoScalingGroupsProperty[];
}

export class Ec2AutoScalingInstanceRunningScheduler extends Construct {
  constructor(scope: Construct, id: string, props: Ec2AutoScalingInstanceRunningSchedulerProps) {
    super(scope, id);

    const account = cdk.Stack.of(this).account;
    //const stackName: string = cdk.Stack.of(this).stackName;
    //const region = cdk.Stack.of(this).region;

    const randomNameKey = crypto.createHash('shake256', { outputLength: 4 })
      .update(`${cdk.Names.uniqueId(scope)}-${cdk.Names.uniqueId(this)}`)
      .digest('hex');

    // ðŸ‘‡EventBridge Scheduler IAM Role
    const schedulerExecutionRole = new iam.Role(this, 'SchedulerExecutionRole', {
      roleName: `ec2-auto-scaling-instance-running-schedule-${randomNameKey}-exec-role`,
      assumedBy: new iam.ServicePrincipal('scheduler.amazonaws.com'),
      inlinePolicies: {
        ['ec2-auto-scaling-desired-capacity-policy']: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                'autoscaling:SetDesiredCapacity',
              ],
              resources: [
                `arn:aws:autoscaling:*:${account}:autoScalingGroup:*:autoScalingGroupName/*`,
              ],
            }),
          ],
        }),
      },
    });

    for (const target of props.targets) {

      // ðŸ‘‡Create random string
      const scheduleNameKey = crypto.createHash('shake256', { outputLength: 4 })
        .update(target.groupName)
        .digest('hex');

      // ðŸ‘‡EventBridge Scheduler for DB Instance Stop
      new scheduler.CfnSchedule(this, `StopSchedule${scheduleNameKey.toUpperCase()}`, {
        name: `auto-stop-ec2-auto-scaling-instance-${scheduleNameKey}-schedule`,
        description: `auto stop ec2 auto scaling group (${target.groupName}) schedule.`,
        state: 'ENABLED',
        //groupName: scheduleGroup.name, // default
        flexibleTimeWindow: {
          mode: 'OFF',
        },
        scheduleExpressionTimezone: target.stopSchedule.timezone,
        scheduleExpression: (() => {
          // default: weekday 21:10
          const minute: string = target.stopSchedule.minute ?? '10';
          const hour: string = target.stopSchedule.hour ?? '21';
          const week: string = target.stopSchedule.week ?? 'MON-FRI';
          return `cron(${minute} ${hour} ? * ${week} *)`;
        })(),
        target: {
          arn: 'arn:aws:scheduler:::aws-sdk:autoscaling:setDesiredCapacity',
          roleArn: schedulerExecutionRole.roleArn,
          input: JSON.stringify({
            AutoScalingGroupName: target.groupName,
            DesiredCapacity: 0,
          }),
          retryPolicy: {
            maximumEventAgeInSeconds: 60,
            maximumRetryAttempts: 0,
          },
        },
      });

      // ðŸ‘‡EventBridge Scheduler for DB Instance Start
      new scheduler.CfnSchedule(this, `StartSchedule${scheduleNameKey.toUpperCase()}`, {
        name: `auto-start-ec2-auto-scaling-instance-${scheduleNameKey}-schedule`,
        description: `auto start ec2 auto scaling group (${target.groupName}) schedule.`,
        state: 'ENABLED',
        //groupName: scheduleGroup.name, // default
        flexibleTimeWindow: {
          mode: 'OFF',
        },
        scheduleExpressionTimezone: target.startSchedule.timezone,
        scheduleExpression: (() => {
          // default: weekday 07:50
          const minute: string = target.startSchedule.minute ?? '50';
          const hour: string = target.startSchedule.hour ?? '7';
          const week: string = target.startSchedule.week ?? 'MON-FRI';
          return `cron(${minute} ${hour} ? * ${week} *)`;
        })(),
        target: {
          arn: 'arn:aws:scheduler:::aws-sdk:autoscaling:setDesiredCapacity',
          roleArn: schedulerExecutionRole.roleArn,
          input: JSON.stringify({
            AutoScalingGroupName: target.groupName,
            DesiredCapacity: target.runningDesiredCapacity,
          }),
          retryPolicy: {
            maximumEventAgeInSeconds: 60,
            maximumRetryAttempts: 0,
          },
        },
      });
    }
  }
}
