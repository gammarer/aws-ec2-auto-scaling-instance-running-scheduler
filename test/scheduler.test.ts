import { App, Stack } from 'aws-cdk-lib';
import { Template, Match } from 'aws-cdk-lib/assertions';
import { Ec2AutoScalingInstanceRunningScheduler } from '../src';

describe('Ec2AutoScalingInstanceRunningScheduler Testing', () => {

  describe('default schedule', () => {
    const app = new App();
    const stack = new Stack(app, 'TestingStack', {
      env: {
        account: '123456789012',
        region: 'us-east-1',
      },
    });

    new Ec2AutoScalingInstanceRunningScheduler(stack, 'Ec2AutoScalingInstanceRunningScheduler', {
      targets: [
        {
          groupName: 'example-scaling-group',
          runningDesiredCapacity: 2,
          startSchedule: {
            timezone: 'UTC',
          },
          stopSchedule: {
            timezone: 'UTC',
          },
        },
      ],
    });

    const template = Template.fromStack(stack);

    // Schedule execution role
    it('Should have schedule execution role', async () => {
      template.hasResourceProperties('AWS::IAM::Role', Match.objectEquals({
        RoleName: Match.stringLikeRegexp('ec2-auto-scaling-instance-running-schedule-.*-exec-role'),
        AssumeRolePolicyDocument: Match.objectEquals({
          Version: '2012-10-17',
          Statement: Match.arrayWith([
            Match.objectEquals({
              Effect: 'Allow',
              Principal: {
                Service: 'scheduler.amazonaws.com',
              },
              Action: 'sts:AssumeRole',
            }),
          ]),
        }),
        Policies: Match.arrayEquals([
          {
            PolicyName: 'ec2-auto-scaling-desired-capacity-policy',
            PolicyDocument: Match.objectEquals({
              Version: '2012-10-17',
              Statement: [
                Match.objectEquals({
                  Effect: 'Allow',
                  Action: 'autoscaling:SetDesiredCapacity',
                  Resource: 'arn:aws:autoscaling:*:123456789012:autoScalingGroup:*:autoScalingGroupName/*',
                }),
              ],
            }),
          },
        ]),
      }));
    });

    // Start Schedule testing
    it('Should have Start Schedule', async () => {
      template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
        Name: Match.stringLikeRegexp('auto-start-ec2-auto-scaling-instance-.*-schedule'),
        Description: Match.anyValue(),
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        ScheduleExpressionTimezone: 'UTC',
        ScheduleExpression: 'cron(50 7 ? * MON-FRI *)',
        Target: Match.objectEquals({
          Arn: 'arn:aws:scheduler:::aws-sdk:autoscaling:setDesiredCapacity',
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('Ec2AutoScalingInstanceRunningSchedulerSchedulerExecutionRole.*'),
              'Arn',
            ],
          },
          Input: JSON.stringify({
            AutoScalingGroupName: 'example-scaling-group',
            DesiredCapacity: 2,
          }),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        }),
      }));
    });

    // Stop Schedule testing
    it('Should have Stop Schedule', async () => {
      template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
        Name: Match.stringLikeRegexp('auto-stop-ec2-auto-scaling-instance-.*-schedule'),
        Description: Match.anyValue(),
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        ScheduleExpressionTimezone: 'UTC',
        ScheduleExpression: 'cron(10 21 ? * MON-FRI *)',
        Target: Match.objectEquals({
          Arn: 'arn:aws:scheduler:::aws-sdk:autoscaling:setDesiredCapacity',
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('Ec2AutoScalingInstanceRunningSchedulerSchedulerExecutionRole.*'),
              'Arn',
            ],
          },
          Input: JSON.stringify({
            AutoScalingGroupName: 'example-scaling-group',
            DesiredCapacity: 0,
          }),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        }),
      }));
    });

    it('Should match snapshot', async () => {
      expect(template.toJSON()).toMatchSnapshot('default');
    });

  });

  describe('specify schedule', () => {
    const app = new App();
    const stack = new Stack(app, 'TestingStack', {
      env: {
        account: '123456789012',
        region: 'us-east-1',
      },
    });

    new Ec2AutoScalingInstanceRunningScheduler(stack, 'Ec2AutoScalingInstanceRunningScheduler', {
      targets: [
        {
          groupName: 'example-scaling-group',
          runningDesiredCapacity: 2,
          startSchedule: {
            timezone: 'Asia/Tokyo',
            minute: '55',
            hour: '8',
            week: 'MON-FRI',
          },
          stopSchedule: {
            timezone: 'Asia/Tokyo',
            minute: '5',
            hour: '19',
            week: 'MON-FRI',
          },
        },
      ],
    });

    const template = Template.fromStack(stack);

    // Start Schedule testing
    it('Should have Start Schedule', async () => {
      template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
        Name: Match.stringLikeRegexp('auto-start-ec2-auto-scaling-instance-.*-schedule'),
        Description: Match.anyValue(),
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        ScheduleExpressionTimezone: 'Asia/Tokyo',
        ScheduleExpression: 'cron(55 8 ? * MON-FRI *)',
        Target: Match.objectEquals({
          Arn: 'arn:aws:scheduler:::aws-sdk:autoscaling:setDesiredCapacity',
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('Ec2AutoScalingInstanceRunningSchedulerSchedulerExecutionRole.*'),
              'Arn',
            ],
          },
          Input: JSON.stringify({
            AutoScalingGroupName: 'example-scaling-group',
            DesiredCapacity: 2,
          }),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        }),
      }));
    });

    // Stop Schedule testing
    it('Should have Stop Schedule', async () => {
      template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
        Name: Match.stringLikeRegexp('auto-stop-ec2-auto-scaling-instance-.*-schedule'),
        Description: Match.anyValue(),
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        ScheduleExpressionTimezone: 'Asia/Tokyo',
        ScheduleExpression: 'cron(5 19 ? * MON-FRI *)',
        Target: Match.objectEquals({
          Arn: 'arn:aws:scheduler:::aws-sdk:autoscaling:setDesiredCapacity',
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('Ec2AutoScalingInstanceRunningSchedulerSchedulerExecutionRole.*'),
              'Arn',
            ],
          },
          Input: JSON.stringify({
            AutoScalingGroupName: 'example-scaling-group',
            DesiredCapacity: 0,
          }),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        }),
      }));
    });

    it('Should match snapshot', async () => {
      expect(template.toJSON()).toMatchSnapshot('specify');
    });

  });
});