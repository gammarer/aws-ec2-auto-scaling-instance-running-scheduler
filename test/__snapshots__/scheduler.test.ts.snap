// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ec2AutoScalingInstanceRunningScheduler Testing default schedule Should match snapshot: default 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Ec2AutoScalingInstanceRunningSchedulerSchedulerExecutionRole7672084B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "scheduler.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "autoscaling:SetDesiredCapacity",
                  "Effect": "Allow",
                  "Resource": "arn:aws:autoscaling:*:123456789012:autoScalingGroup:*:autoScalingGroupName/*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "ec2-auto-scaling-desired-capacity-policy",
          },
        ],
        "RoleName": "ec2-auto-scaling-instance-running-schedule-ce7ae6a9-exec-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "Ec2AutoScalingInstanceRunningSchedulerStartSchedule84CCB259C47E963F": {
      "Properties": {
        "Description": "auto start ec2 auto scaling group (example-scaling-group) schedule.",
        "FlexibleTimeWindow": {
          "Mode": "OFF",
        },
        "Name": "auto-start-ec2-auto-scaling-instance-84ccb259-schedule",
        "ScheduleExpression": "cron(50 7 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "UTC",
        "State": "ENABLED",
        "Target": {
          "Arn": "arn:aws:scheduler:::aws-sdk:autoscaling:setDesiredCapacity",
          "Input": "{"AutoScalingGroupName":"example-scaling-group","DesiredCapacity":2}",
          "RetryPolicy": {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": {
            "Fn::GetAtt": [
              "Ec2AutoScalingInstanceRunningSchedulerSchedulerExecutionRole7672084B",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
    "Ec2AutoScalingInstanceRunningSchedulerStopSchedule84CCB25922932086": {
      "Properties": {
        "Description": "auto stop ec2 auto scaling group (example-scaling-group) schedule.",
        "FlexibleTimeWindow": {
          "Mode": "OFF",
        },
        "Name": "auto-stop-ec2-auto-scaling-instance-84ccb259-schedule",
        "ScheduleExpression": "cron(10 21 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "UTC",
        "State": "ENABLED",
        "Target": {
          "Arn": "arn:aws:scheduler:::aws-sdk:autoscaling:setDesiredCapacity",
          "Input": "{"AutoScalingGroupName":"example-scaling-group","DesiredCapacity":0}",
          "RetryPolicy": {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": {
            "Fn::GetAtt": [
              "Ec2AutoScalingInstanceRunningSchedulerSchedulerExecutionRole7672084B",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Ec2AutoScalingInstanceRunningScheduler Testing specify schedule Should match snapshot: specify 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Ec2AutoScalingInstanceRunningSchedulerSchedulerExecutionRole7672084B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "scheduler.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "autoscaling:SetDesiredCapacity",
                  "Effect": "Allow",
                  "Resource": "arn:aws:autoscaling:*:123456789012:autoScalingGroup:*:autoScalingGroupName/*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "ec2-auto-scaling-desired-capacity-policy",
          },
        ],
        "RoleName": "ec2-auto-scaling-instance-running-schedule-ce7ae6a9-exec-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "Ec2AutoScalingInstanceRunningSchedulerStartSchedule84CCB259C47E963F": {
      "Properties": {
        "Description": "auto start ec2 auto scaling group (example-scaling-group) schedule.",
        "FlexibleTimeWindow": {
          "Mode": "OFF",
        },
        "Name": "auto-start-ec2-auto-scaling-instance-84ccb259-schedule",
        "ScheduleExpression": "cron(55 8 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "Asia/Tokyo",
        "State": "ENABLED",
        "Target": {
          "Arn": "arn:aws:scheduler:::aws-sdk:autoscaling:setDesiredCapacity",
          "Input": "{"AutoScalingGroupName":"example-scaling-group","DesiredCapacity":2}",
          "RetryPolicy": {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": {
            "Fn::GetAtt": [
              "Ec2AutoScalingInstanceRunningSchedulerSchedulerExecutionRole7672084B",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
    "Ec2AutoScalingInstanceRunningSchedulerStopSchedule84CCB25922932086": {
      "Properties": {
        "Description": "auto stop ec2 auto scaling group (example-scaling-group) schedule.",
        "FlexibleTimeWindow": {
          "Mode": "OFF",
        },
        "Name": "auto-stop-ec2-auto-scaling-instance-84ccb259-schedule",
        "ScheduleExpression": "cron(5 19 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "Asia/Tokyo",
        "State": "ENABLED",
        "Target": {
          "Arn": "arn:aws:scheduler:::aws-sdk:autoscaling:setDesiredCapacity",
          "Input": "{"AutoScalingGroupName":"example-scaling-group","DesiredCapacity":0}",
          "RetryPolicy": {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": {
            "Fn::GetAtt": [
              "Ec2AutoScalingInstanceRunningSchedulerSchedulerExecutionRole7672084B",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
